### PR TITLE
Remove HMP from optimum-habana

### PR DESCRIPTION
**Pull request by jwieczorekhabana**
HMP is deprecated in favor of PyTorch autocast
- removed hmp usage
- removed setting autocast env variables through GaudiConfig
- updated tests
- updated docs
- updated README.md files

### CODE REVIEW

**Comment by regisss** [comment]
@jwieczorekhabana Thanks for opening this PR! I'll review it in the beginning of next week and I'll try to accelerate on merging the autocast PRs.

File: optimum/habana/transformers/gaudi_configuration.py
Comment (regisss) [inline optimum/habana/transformers/gaudi_configuration.py:91]: We still need this method to be able to specify autocast custom op lists in a Gaudi config

File: tests/test_gaudi_configuration.py
Comment (regisss) [inline tests/test_gaudi_configuration.py:85]: We should keep this test as `write_bf16_fp32_ops_to_text_files` will still be used to specify custom op lists.
I propose to instantiate a Gaudi config such as:
```py
gaudi_config = GaudiConfig(
    autocast_bf16_ops=[
        "add",
        "addmm",
        "bmm",
        "div",
        "dropout",
        "gelu",
        "iadd",
        "linear",
        "layer_norm",
        "matmul",
        "mm",
        "rsub",
        "softmax",
        "truediv",
    ],
    autocast_fp32_ops=[
        "embedding",
        "nll_loss",
        "log_softmax",
    ],
)
```

File: docs/source/package_reference/gaudi_config.mdx
Comment (regisss) [inline docs/source/package_reference/gaudi_config.mdx:45]: I think we should mention:
- `use_torch_autocast` but saying that `--bf16` should be favored as `use_torch_autocast` is used to define a good pre-defined config
- `autocast_bf16_ops` and `autocast_fp32_ops` as #308 enables users to specify cutom op lists but saying that the default should work for most models

File: docs/source/usage_guides/accelerate_training.mdx
Comment (regisss) [inline docs/source/usage_guides/accelerate_training.mdx:60]:
```suggestion
Optimum Habana enables mixed precision training in a similar fashion as 🤗 Transformers.
```

File: docs/source/usage_guides/accelerate_training.mdx
Comment (regisss) [inline docs/source/usage_guides/accelerate_training.mdx:66]: I would keep this

File: docs/source/usage_guides/accelerate_training.mdx
Comment (regisss) [inline docs/source/usage_guides/accelerate_training.mdx:65]:
```suggestion
<Tip>
```
I think this one doesn't have to be a warning, it's rather extra information

File: docs/source/usage_guides/accelerate_training.mdx
Comment (regisss) [inline docs/source/usage_guides/accelerate_training.mdx:93]: I would still keep a part of this to show how to specify custom op lists. We can add a link to the GPT2 Gaudi config when it is updated.

**Review by regisss — COMMENTED** [review]
@jwieczorekhabana I left several comments, the main point being that #308 enables users to specify their own op lists in Gaudi config files so this should be documented, tested, etc (which should basically consist in changing a few variable names).

For OPT, I need to check if we need to disable autocast for this specific line.

**Note that we should not merge this PR before we manage to make Wav2Vec2 work with autocast.** It is being investigated but no fix has been merged yet.

File: docs/source/usage_guides/accelerate_training.mdx
Comment (jwieczorekhabana) [reply docs/source/usage_guides/accelerate_training.mdx:66]: But those operators are incompatible with autocast. HMP and autocast operate on different software levels. Please see: https://docs.habana.ai/en/latest/PyTorch/PyTorch_Mixed_Precision/Autocast.html#override-options

File: docs/source/usage_guides/accelerate_training.mdx
Comment (jwieczorekhabana) [reply docs/source/usage_guides/accelerate_training.mdx:65]: I'd rather keep it as a warning as the names of operations in HMP and autocast may differ. Wouldn't it be important for users switching from HMP to autocast.

File: docs/source/usage_guides/accelerate_training.mdx
Comment (jwieczorekhabana) [reply docs/source/usage_guides/accelerate_training.mdx:93]: But shouldn't users provide custom lists in a similar way to other training demos outside of HuggingFace? We can keep those in GaudiConfig to make sure they are optimized for specific model.

File: docs/source/usage_guides/accelerate_training.mdx
Comment (regisss) [reply docs/source/usage_guides/accelerate_training.mdx:66]: No problem, we don't have to keep the same operators. Maybe it will just be easier to refer to GPT2's Gaudi config.

File: docs/source/package_reference/gaudi_config.mdx
Comment (regisss) [reply docs/source/package_reference/gaudi_config.mdx:45]: As discussed by email, regarding `autocast_bf16_ops` and `autocast_fp32_ops`, I'm fine with saying that the env variable way should be favored. But they should still be documented.

File: docs/source/package_reference/gaudi_config.mdx
Comment (regisss) [inline docs/source/package_reference/gaudi_config.mdx:24]: You can specify here that setting the env variables `LOWER_LIST` and `FP32_LIST` is the prioritized way of specifying these ops

File: docs/source/usage_guides/accelerate_training.mdx
Comment (regisss) [inline docs/source/usage_guides/accelerate_training.mdx:60]:
```suggestion
Optimum Habana enables mixed precision training in a similar fasion as 🤗 Transofrmers:
```

File: tests/test_gaudi_configuration.py
Comment (regisss) [inline tests/test_gaudi_configuration.py:26]: We still need those in `test_write_bf16_fp32_ops_to_text_files`

File: optimum/habana/transformers/models/gpt2/modeling_gpt2.py
Comment (regisss) [inline optimum/habana/transformers/models/gpt2/modeling_gpt2.py:334]: Let's keep this line and change "HMP" by "autocast"

File: optimum/habana/diffusers/pipelines/pipeline_utils.py
Comment (regisss) [inline optimum/habana/diffusers/pipelines/pipeline_utils.py:121]: We need to keep these two lines for autocast as `bf16_full_eval` doesn't rely on autocast (it basically converts everything into bf16)

**Review by regisss — COMMENTED** [review]
Thanks for updating the PR! I left a few comments.

Could you also run the following to make the code style check pass please?
```
pip install black ruff
make style
```

File: docs/source/usage_guides/accelerate_training.mdx
Comment (jwieczorekhabana) [reply docs/source/usage_guides/accelerate_training.mdx]: done

File: tests/test_gaudi_configuration.py
Comment (jwieczorekhabana) [reply tests/test_gaudi_configuration.py:26]: done

File: optimum/habana/diffusers/pipelines/pipeline_utils.py
Comment (jwieczorekhabana) [reply optimum/habana/diffusers/pipelines/pipeline_utils.py:121]: done

File: docs/source/usage_guides/accelerate_training.mdx
Comment (regisss) [inline docs/source/usage_guides/accelerate_training.mdx:60]:
```suggestion
Optimum Habana enables mixed precision training in a similar fashion as 🤗 Transformers:
```

**Review by regisss — APPROVED** [review]
LGTM!

Let's wait to know what happens with Wav2Vec2 training before merging anything cc @libinta

**Comment by regisss** [comment]
@jwieczorekhabana will the Gaudi config for Wav2Vec2 be the following?
```json
{
  "use_fused_adam": true,
  "use_fused_clip_norm": true,
  "use_torch_autocast": true
}
```
Or will there be any specific bf16 ops?

**Comment by jwieczorekhabana** [comment]
> @jwieczorekhabana will the Gaudi config for Wav2Vec2 be the following?
> 
> ```json
> {
>   "use_fused_adam": true,
>   "use_fused_clip_norm": true,
>   "use_torch_autocast": true
> }
> ```
> 
> Or will there be any specific bf16 ops?

That'll be it. There is a hanging PR on HuggingFace https://huggingface.co/Habana/wav2vec2/discussions/2/files that can be merged after 1.13 release.
Similarly for gpt2 there is a closed PR that can be reopened https://huggingface.co/Habana/gpt2/discussions/4/files

Release 1.13 addresses issues both of those topologies had.

**Comment by regisss** [comment]
Cool, I'll merge them when 1.13 is released then :+1:

**Comment by regisss** [comment]
@jwieczorekhabana Can you rebase this branch on main? There are merge conflicts in `tests/test_diffusers.py` so GitHub won't let me merge the PR when 1.13 is released.

**Comment by jwieczorekhabana** [comment]
> @jwieczorekhabana Can you rebase this branch on main? There are merge conflicts in `tests/test_diffusers.py` so GitHub won't let me merge the PR when 1.13 is released.

@regisss Done

**Comment by regisss** [comment]
@jwieczorekhabana It seems there are still some merge conflicts in `tests/test_diffusers.py`

**Comment by regisss** [comment]
@jwieczorekhabana I pushed a few commits to solve the CI that was not passing

**Review by regisss — APPROVED** [review]

**Pull request merged**
